cheatsheets/Java_Security_Cheat_Sheet.md

```diff
@@ -0,0 +1,971 @@
+        </triggeringPolicy>
+        <encoder>
+            <!-- Encode any CRLF chars in the message and limit
+                 its maximum size to 500 characters -->
+            <pattern>%relative [%thread] %-5level %logger{35} - %crlf(%.-500msg) %n</pattern>
+        </encoder>
+    </appender>
+    <root level="debug">
+        <appender-ref ref="RollingFile" />
+    </root>
+</configuration>
+```
+
+You also have to add the [OWASP Security Logging](https://github.com/augustd/owasp-security-logging/wiki/Usage-with-Logback) dependency to your project.
+
+Usage of the logger at code level:
+
+``` java
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+...
+// No special action needed because security actions
+// are performed at the logging policy level
+Logger logger = LoggerFactory.getLogger(MyClass.class);
+logger.info(logMessage);
+...
+```
+
+#### References
+
+- [PatternLayout](https://logging.apache.org/log4j/2.x/manual/layouts.html#PatternLayout) (See the `encode{}{CRLF}` function)
+
+```text
+Note that the default Log4j2 encode{} encoder is HTML, which does NOT prevent log injection.
+
+It prevents XSS attacks against viewing logs using a browser.
+
+OWASP recommends defending against XSS attacks in such situations in the log viewer application itself,
+not by preencoding all the log messages with HTML encoding as such log entries may be used/viewed in many
+other log viewing/analysis tools that don't expect the log data to be pre-HTML encoded.
+```
+
+- [LOG4J Configuration](https://logging.apache.org/log4j/2.x/manual/configuration.html)
+- [LOG4J Appender](https://logging.apache.org/log4j/2.x/manual/appenders.html)
+- [Log Forging](https://github.com/javabeanz/owasp-security-logging/wiki/Log-Forging) - See the Logback section about the `CRLFConverter` this library provides.
+- [Usage of OWASP Security Logging with Logback](https://github.com/javabeanz/owasp-security-logging/wiki/Usage-with-Logback)
+
+## Cryptography
+
+### General cryptography guidance
+
+- **Never, ever write your own cryptographic functions.**
+- Wherever possible, try and avoid writing any cryptographic code at all. Instead try and either use pre-existing secret management solutions or the secret management solution provided by your cloud provider. For more information, see the [OWASP Secrets Management Cheat Sheet](/Secrets_Management_Cheat_Sheet).
+- Make sure your application or protocol can easily support a future change of cryptographic algorithms.
+- Use your package manager wherever possible to keep all of your packages up to date. Watch the updates on your development setup, and plan updates to your applications accordingly.
+- Wherever possible, use a trusted and well known implementation library rather than using the libraries built into JCA/JCE as it is far too easy to make cryptographic errors with them.
+- We will show examples below based on Google Tink which is an opinionated library for performing encryption operations.

```

Comment:
I'm not sure what you mean by "opinionated library"? A inanimate object like a library, can't really hold an opinion, and library developers almost always do. (I personally have more opinions than I can remember!)

But my guess is that you are trying to say something else here. Give us a hint or use a thesaurus.

Reply:
Added a link for those who are unclear